File: January2021/CPP/Week1/Remove Duplicates from Sorted List II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    //TC:O(N)| SC:O(1)
    ListNode *deleteDuplicates(ListNode *head)
    {
        ListNode *dummyNode = new ListNode(INT_MIN);

        dummyNode->next = head;
        ListNode *curr = head;
        ListNode *prev = dummyNode;

        while (curr != NULL && curr->next != NULL)
        {

            bool duplication = false;
            while (curr != NULL && curr->next != NULL && curr->val == curr->next->val)
            {
                duplication = true;
                curr = curr->next;
            }

            if (!duplication)
            {
                prev = prev->next;
            }
            else
            {
                prev->next = curr->next;
            }
            curr = curr->next;
        }

        return dummyNode->next;
    }
};

File: March2022/Java/RemoveDuplicatesFromSortedListII.java
class Solution {

    // TC : O(n)
    // SC : O(1)
    public ListNode deleteDuplicates(ListNode head) {
        if(head == null || head.next == null){
            return head;
        }

        ListNode dummy = new ListNode(Integer.MIN_VALUE);
        dummy.next = head;

        ListNode it = head;
        ListNode prev = dummy;

        while(it!=null && it.next!=null){

            boolean foundDuplication = false;
            while(it!=null && it.next!=null && it.val == it.next.val){
                foundDuplication  = true;
                it = it.next;
            }

            if(!foundDuplication){
                prev = prev.next;
            } else {
                prev.next = it.next;
            }

            it = it.next;
        }
        return dummy.next;
    }
}